devito/types/petsc.py

```diff
@@ -0,0 +1,80 @@
+from devito.tools import dtype_to_petsctype, CustomDtype
+from devito.types import LocalObject
+from devito.types.array import ArrayBasic
+import numpy as np
+
+
+class DM(LocalObject):
+    """
+    PETSc Data Management object (DM).
+    """
+    dtype = CustomDtype('DM')
+
+
+class Mat(LocalObject):
+    """
+    PETSc Matrix object (Mat).
+    """
+    dtype = CustomDtype('Mat')
+
+
+class Vec(LocalObject):
+    """
+    PETSc Vector object (Vec).
+    """
+    dtype = CustomDtype('Vec')
+
+
+class PetscMPIInt(LocalObject):
+    """
+    PETSc datatype used to represent `int` parameters
+    to MPI functions.
+    """
+    dtype = CustomDtype('PetscMPIInt')
+
+
+class KSP(LocalObject):
+    """
+    PETSc KSP : Linear Systems Solvers.
+    Manages Krylov Methods.
+    """
+    dtype = CustomDtype('KSP')
+
+
+class PC(LocalObject):
+    """
+    PETSc object that manages all preconditioners (PC).
+    """
+    dtype = CustomDtype('PC')
+
+
+class KSPConvergedReason(LocalObject):
+    """
+    PETSc object - reason a Krylov method was determined
+    to have converged or diverged.
+    """
+    dtype = CustomDtype('KSPConvergedReason')
+
+
+class PETScArray(ArrayBasic):
+    """
+    PETScArrays are generated by the compiler only and represent
+    a customised variant of ArrayBasic. They are designed to
+    avoid generating a cast in the low-level code.
+    """
+
+    _data_alignment = False
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return kwargs.get('dtype', np.float32)
+
+    @property
+    def _C_ctype(self):
+        petsc_type = dtype_to_petsctype(self.dtype)
```

Comment:
nitpicking:

PETSc is so special that I would maybe put `dtype_to_petsctype` in here rather than in `tools/`